torch/_inductor/codegen/wrapper.py

```diff
@@ -897,10 +954,163 @@ def wrap_kernel_call(self, name, call_args):
+
+            CUfunction loadKernel(const std::string &filePath, const std::string &funcName) {
+                CUmodule mod;
+                CUfunction func;
+                AT_CUDA_DRIVER_CHECK(cuModuleLoad(&mod, filePath.c_str()));
+                AT_CUDA_DRIVER_CHECK(cuModuleGetFunction(&func, mod, funcName.c_str()));
+                return func;
+            }
+
+            void launchKernel(
+                    CUfunction func,
+                    int gridX,
+                    int gridY,
+                    int gridZ,
+                    int numWraps,
+                    int sharedMemBytes,
+                    void* args[],
+                    int device_index) {
+                at::cuda::CUDAGuard device_guard(device_index);
+                AT_CUDA_DRIVER_CHECK(cuLaunchKernel(
+                    func, gridX, gridY, gridZ, 32*numWraps, 1, 1, sharedMemBytes,
+                    at::cuda::getCurrentCUDAStream(), args, nullptr));
+            }
+            """
+        )
+
+    def define_kernel(self, name: str, kernel: str, kernel_path: str = None):
+        self.kernels[kernel] = name
+
+    def generate(self):
+        self.prefix.writeline("\n")
+        for kernel in self.kernels.values():
+            self.prefix.writeline(f"static CUfunction {kernel} = nullptr;")
+        self.prefix.writeline("\n")
+        return super().generate()
+
+    def generate_load_kernel(self, name: str = None):
+        self.writeline(f"if ({name} == nullptr) {{")
+        self.writeline(
+            DeferredKernelNameLine(
+                name, f"""     {name} = loadKernel("{name}.cubin", "MANGLED_NAME");"""
+            )
+        )
+        self.writeline("}")
+
+    def generate_args_decl(self, call_args):
+        # TODO: only works for constant now, need type info
+        new_args = []
+        for arg in call_args:
+            var_name = f"var_{self.arg_var_id}"
+            if is_int(arg):
+                self.writeline(f"int {var_name} = {arg};")
+            elif is_float(arg):
+                self.writeline(f"float {var_name} = {arg};")
+            else:
+                self.writeline(
+                    f"CUdeviceptr {var_name} = reinterpret_cast<CUdeviceptr>({arg}.data_ptr());"
+                )
+            new_args.append(f"&{var_name}")
+            self.arg_var_id += 1
+
+        return ", ".join(new_args)
+
+    def generate_kernel_call(self, name, call_args, device_index):
+        self.generate_load_kernel(name)
+        call_args = self.generate_args_decl(call_args)
+        self.writeline(
+            f"void* kernel_args_{self.kernel_callsite_id}[] = {{{call_args}}};"
+        )
+        self.writeline(
+            DeferredKernelParamLine(
+                name,
+                f"launchKernel({name}, LAUNCH_PARAMS, kernel_args_{self.kernel_callsite_id}, {device_index});",
+            )
+        )
+        self.kernel_callsite_id += 1
+
+
+class TwoPassAotWrapperCodeGen(WrapperCodeGen):
```

Comment:
This two pass approach feels a bit clunky when implemented here.  I wonder if we could simplify things by moving the two passes up to the caller's level.

Basically, require the user to do a warmup run.  Or have a wrapper function outside inductor that does the warmup run via two calls to `compile_fx`.

When compiling in AOT mode, we can basically just assert that everything is a cache hit, so all the autotuning will be a no-op.

Reply:
Requiring user to do an explicit warmup run sounds a little award to me, although I agree calling `compile_fx` twice makes the code cleaner and more robust. I plan to change `aot_compile` in `_inductor/__init__.py` to achieve that. How does that sound?